src/include/simeng/kernel/Process.hh

```diff
@@ -0,0 +1,157 @@
+ * |               |
+ * |  ELF-defined  |
+ * | process image |
+ * |               |
+ * |---------------| <- 0x0
+ *
+ */
+class Process {
+ public:
+  /** Construct a SimOS Process from a vector of command-line arguments.
+   *
+   * The first argument is a path to an executable ELF file. */
+  Process(const std::vector<std::string>& commandLine,
+          std::shared_ptr<simeng::memory::Mem> memory,
+          std::vector<RegisterFileStructure> regFileStructure, uint64_t TGID,
+          uint64_t TID);
+
+  /** Construct a SimOS Process from region of instruction memory, with the
+   * entry point fixed at 0. */
+  Process(span<char> instructions, std::shared_ptr<simeng::memory::Mem> memory,
+          std::vector<RegisterFileStructure> regFileStructure, uint64_t TGID,
+          uint64_t TID);
+
+  ~Process();
+
+  /** Get the address of the start of the heap region. */
+  uint64_t getHeapStart() const;
+
+  /** Get the address of the top of the stack. */
+  uint64_t getStackStart() const;
+
+  /** Get the address of the start of the mmap region. */
+  uint64_t getMmapStart() const;
+
+  /** Get the page size. */
+  uint64_t getPageSize() const;
+
+  /** Get the size of the process image. */
+  uint64_t getProcessImageSize() const;
+
+  /** Get the entry point. */
+  uint64_t getEntryPoint() const;
+
+  /** Get the initial stack pointer address. */
+  uint64_t getStackPointer() const;
+
+  /** Get the path of the executable. */
+  std::string getPath() const;
+
+  /** Get the memory region for this process. */
+  MemRegion& getMemRegion() { return memRegion_; }
+
+  /** Check whether the process image was created successfully. */
+  bool isValid() const;
+
+  /** Get the process' TGID. */
+  uint64_t getTGID() const { return TGID_; }
+
+  /** Get the process' TID. */
+  uint64_t getTID() const { return TID_; }
+
+  /** The virtual file descriptor mapping table. */
+  std::vector<int64_t> fileDescriptorTable_;
+
+  /** Set of deallocated virtual file descriptors available for reuse. */
+  std::set<int64_t> freeFileDescriptors_;
+
+  // Thread state
+  // TODO: Support multiple threads per process
+  /** The clear_child_tid value. */
+  uint64_t clearChildTid = 0;
+
+  /** Current status of the process. */
+  procStatus status_ = procStatus::waiting;
+
+  cpuContext context_;
+
+ private:
+  /** MemRegion of the Process Image. */
+  MemRegion memRegion_;
+
+  /** Create and populate the initial process stack and returns the stack
+   * pointer. */
+  uint64_t createStack(char** processImage, uint64_t stackStart);
+
+  /** The entry point of the process. */
+  uint64_t entryPoint_ = 0;
+
+  /** The process command and its arguments. */
+  std::vector<std::string> commandLine_;
+
+  /** Whether the process image was created successfully. */
+  bool isValid_ = false;
+
+  /** The process' Thread Group ID, exactly equivalent to its Process ID (PID).
+   */
+  uint64_t TGID_;
+
+  /** The process' Thread ID, its globally unique identifier.
```

Comment:
I am unsure what to re-word this to. I think "The globally unique process Thread ID." is more misleading as it sounds like the process is the globally unique item here, rather than the Thread ID. 

Perhaps "The process' Thread ID - a globally unique identifyer."?

Reply:
Fine with that